proof/invariant-abstract/Tcb_AI.thy

```diff
@@ -1654,68 +1650,159 @@ lemma mapM_priorities:
-  apply (wpsimp simp: reorder_ntfn_def live_def live_ntfn_def
-                  wp: set_ntfn_minor_invs sort_queue_valid_ntfn_rv get_simple_ko_wp)
-  apply (clarsimp simp: get_ntfn_queue_def split: ntfn.splits)
-  apply (rule conjI, clarsimp simp: obj_at_def)
-  apply (rule obj_at_valid_objsE, assumption, fastforce)
-  apply (frule if_live_then_nonz_capD[OF invs_iflive], assumption)
-  by (auto dest!: not_idle_tcb_in_waitingntfn
-            simp: valid_obj_def live_def live_ntfn_def)
+  "\<lbrace>invs and st_tcb_at (\<lambda>st. ntfn_blocked st = Some nptr) tptr\<rbrace> reorder_ntfn nptr tptr \<lbrace>\<lambda>rv. invs\<rbrace>"
+  apply (simp only: reorder_ntfn_def)
+  apply (subst bind_assoc[symmetric, where m="tcb_ep_dequeue tptr _"])
+  apply (rule hoare_seq_ext)+
+     apply (wpsimp wp: set_ntfn_minor_invs get_simple_ko_wp tcb_ep_dequeue_rv_wf'
+                       tcb_ep_dequeue_append_valid_ntfn_rv hoare_vcg_conj_lift
+                 simp: live_def live_ntfn_def pred_conj_def)+
+  apply (frule valid_objs_ko_at[rotated], fastforce)
+  apply (clarsimp simp: valid_obj_def valid_ntfn_def obj_at_def pred_tcb_at_def cong: conj_cong)
+  apply (case_tac "tcb_state tcb"; clarsimp simp: ntfn_blocked_def get_ntfn_queue_def split: ntfn.splits)
+  apply (rule context_conjI)
+   apply (fastforce simp: refs_of_rev obj_at_def get_refs_def
+                    dest: invs_sym_refs sym_refs_ko_atD[rotated]
+                   split: option.splits)
+  apply (intro conjI)
+     apply fastforce
+    apply (fastforce simp: valid_idle_def pred_tcb_at_def obj_at_def dest: invs_valid_idle)
+   apply (erule_tac x="idle_thread s" in ballE; clarsimp?)
+   apply (frule_tac p="nptr" in sym_refs_ko_atD[OF _ invs_sym_refs, rotated])
+    apply (simp add: obj_at_def)
+   apply (clarsimp simp: refs_of_rev obj_at_def is_tcb_def)
+   apply (erule_tac x="(idle_thread s, NTFNSignal)" in ballE; simp)
+   apply (fastforce simp: refs_of_rev valid_idle_def pred_tcb_at_def obj_at_def dest: invs_valid_idle)
+  apply (fastforce simp: if_live_then_nonz_cap_def obj_at_def live_def live_ntfn_def dest: invs_iflive)
+  done
 
 lemma set_ep_minor_invs:
   "\<lbrace>invs and obj_at (\<lambda>ko. refs_of ko = ep_q_refs_of val) ptr
          and valid_ep val
          and (\<lambda>s. \<forall>typ. (idle_thread s, typ) \<notin> ep_q_refs_of val)
          and (\<lambda>s. live (Endpoint val) \<longrightarrow> ex_nonz_cap_to ptr s)\<rbrace>
-     set_endpoint ptr val
+   set_endpoint ptr val
    \<lbrace>\<lambda>rv. invs\<rbrace>"
-  apply (wpsimp simp: invs_def valid_state_def valid_pspace_def
-          wp: valid_irq_node_typ valid_ioports_lift simp_del: fun_upd_apply)
+  apply (wpsimp wp: valid_irq_node_typ valid_ioports_lift
+              simp: invs_def valid_state_def valid_pspace_def)
   apply (clarsimp simp: state_refs_of_def obj_at_def ext elim!: rsubst[where P = sym_refs])
   done
 
+lemma tcb_ep_dequeue_valid_SendEP:
```

Comment:
Very minor, but is it worth moving these `tcb_eq_dequeue` lemmas to be with the similar `tcb_ep_append` lemmas in `IpcDet`?